### PR TITLE
Socint 246

### PR DESCRIPTION
**Pull request by KieranWardle**
# Motivation and Context
We don't know what fields the UI mightt need as part of the UacUpdateDTO in future

# What has changed
Sending all available information to UI in UacUpdateDTO

# How to test?
Tested in dev with the corresponding UI changes and cucumber tests ran

# Links
https://collaborate2.ons.gov.uk/jira/browse/SOCINT-251
https://collaborate2.ons.gov.uk/jira/browse/SOCINT-246

### CODE REVIEW

**Review by SteveScorfield — APPROVED** [review]
Looks fine

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/representation/UniqueAccessCodeDTO.java [hunk 1]
```diff
@@ -1,27 +1,18 @@
...
-  private CaseStatus caseStatus;
   private String qid;
-  private String region;
-  private UUID caseId;
-  private UUID collectionExerciseId;
-  private AddressDTO address;
   private boolean receiptReceived;
   private boolean eqLaunched;
   private int wave;
+  private SurveyLiteDTO surveyLiteDTO;
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/representation/UniqueAccessCodeDTO.java:15]: Using the 'DTO' on the end of the attribute name will mean that that will be what appears in the serialisation ie the JSON that goes to the UI, not etiquette

**Review by philwhiles — CHANGES_REQUESTED** [review]
lose the 'DTO'

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/representation/UniqueAccessCodeDTO.java (hunk 1 above)
Comment (KieranWardle) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/representation/UniqueAccessCodeDTO.java]: Done

**Review by SteveScorfield — APPROVED** [review]
Approved new changes

**Review by rpcatling — CHANGES_REQUESTED** [review]
the first thing that leaps out here is lack of type consistency. OK I know the JIRA is inconsistent , but we need to apply some thought to it,
I believe for DTOs (which are NOT stored in firestore):
- for UUID id's use UUID instead of String
- decide on whether you are going to use Integer or int and be consistent - at present we have a mishmash.
- for date/time we are moving towards Java8 types (unless it has to be stored in firestore, then we don't have that choice).  So use LocalDateTime for a date time.  I know the collection exercise is has names startDate and endDate, but they are actually date-times.

File: src/test/resources/uk/gov/ons/ctp/integration/rhsvc/RHSvcBeanMapperTest.CollectionExercise.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  "reference": "MVP012021",
+  "startDate": "2021-09-17T23:59:59.999Z",
+  "endDate": "2021-09-27T23:59:59.999Z",
+  "metadata": {
+    "numberOfWaves": "3",
+    "waveLength": "2",
+    "cohorts": "3",
+    "cohortSchedule": "7"
+  }
+}
```
Comment (rpcatling) [inline src/test/resources/uk/gov/ons/ctp/integration/rhsvc/RHSvcBeanMapperTest.CollectionExercise.json:14]: newline please

File: src/test/resources/uk/gov/ons/ctp/integration/rhsvc/RHSvcBeanMapperTest.SurveyUpdate.json [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+          "expression": "^07[0-9]{9}$"
+        }
+      ]
+    }
+  ],
+  "metadata": {
+    "ex_e4": true
+  },
+  "sampleDefinitionUrl": "https://raw.githubusercontent.com/ONSdigital/ssdc-shared-events/main/sample/social/0.1.0-DRAFT/social.json"
+}
```
Comment (rpcatling) [inline src/test/resources/uk/gov/ons/ctp/integration/rhsvc/RHSvcBeanMapperTest.SurveyUpdate.json:87]: newline please

File: src/test/resources/uk/gov/ons/ctp/integration/rhsvc/endpoint/UniqueAccessCodeEndpointTest.UniqueAccessCodeDTO.json [hunk 4]
```diff
@@ -2,18 +2,39 @@
...
+			"addressLine1": "1 main street",
+			"addressLine2": "upper upperingham",
+			"addressLine3": "",
+			"townName": "upton",
+			"region": "E",
+			"postcode": "UP103UP",
+			"uprn": "123456"
+		}
+	}
 }
```
Comment (rpcatling) [inline src/test/resources/uk/gov/ons/ctp/integration/rhsvc/endpoint/UniqueAccessCodeEndpointTest.UniqueAccessCodeDTO.json:40]: newline please. Also please format consistently.  We generally have 2-space indent for JSON files.

File: src/test/resources/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImplTest.CollectionExercise.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  "reference": "MVP012021",
+  "startDate": "2021-09-17T23:59:59.999Z",
+  "endDate": "2021-09-27T23:59:59.999Z",
+  "metadata": {
+    "numberOfWaves": "3",
+    "waveLength": "2",
+    "cohorts": "3",
+    "cohortSchedule": "7"
+  }
+}
```
Comment (rpcatling) [inline src/test/resources/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImplTest.CollectionExercise.json:14]: newline please

File: src/test/resources/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImplTest.SurveyUpdate.json [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+          "expression": "^07[0-9]{9}$"
+        }
+      ]
+    }
+  ],
+  "metadata": {
+    "ex_e4": true
+  },
+  "sampleDefinitionUrl": "https://raw.githubusercontent.com/ONSdigital/ssdc-shared-events/main/sample/social/0.1.0-DRAFT/social.json"
+}
```
Comment (rpcatling) [inline src/test/resources/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImplTest.SurveyUpdate.json:87]: newline please

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImpl.java [hunk 7]
```diff
@@ -73,17 +87,18 @@ public UniqueAccessCodeDTO getAndAuthenticateUAC(String uacHash) throws CTPExcep
 
   /** Send UacAuthentication event */
   private void sendUacAuthenticationEvent(UniqueAccessCodeDTO data) throws CTPException {
+    UUID caseId = null;
+    if (data.getCollectionCase() != null) {
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImpl.java:91]: i don't think this should ever be null . We are not catering for unlinked UACs now.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImpl.java [hunk 8]
```diff
@@ -47,20 +51,30 @@ public UniqueAccessCodeDTO getAndAuthenticateUAC(String uacHash) throws CTPExcep
...
+
+          // TODO log if missing?
+          // Will always exist if the case exsits as it can't be saved without them
+          Optional<SurveyUpdate> surveyMatch = dataRepo.readSurvey(caseMatch.get().getSurveyId());
+          Optional<CollectionExercise> collexMatch =
+              dataRepo.readCollectionExercise(caseMatch.get().getCollectionExerciseId());
+
+          data = createUniqueAccessCodeDTO(uacMatch.get(), caseMatch, collexMatch, surveyMatch);
         } else {
           // Case NOT found
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImpl.java:63]: please remove code to do with unlinked UACs . This code can be much simplified

**Review by rpcatling — CHANGES_REQUESTED** [review]
more comments

**Review by philwhiles — CHANGES_REQUESTED** [review]
1. When we get a UAC (and Case, and Survey, and CollEx), the only thing we should report to the UI as NotFound is the UAC itself. The only thing the UI and the user has control over is the UAC after all. As requested in the slack thread yesterday that questioned this, anything else should be a SYSTEM_ERROR.
Survey, Collex and Case should all be there - if they are not, it is an exceptional circumstance, something has gone wrong in event delivery or storage, and is a problem within the service, nothing the user can do will change that.

2. In the same method there is use of createException(...) and throws new ....
Same thing, done differently in the same method.
Be consistent please. Suggest losing createException(message) as for it does not allow the TYPE of exception to be chosen by the caller, and even then it provides no utility over 'throw new etc'

**Pull request merged**